controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -453,107 +443,185 @@ public String getScopeName() {
+                                        .anyMatch(y -> y.getKey().equals(x.getKey())
+                                                && y.getValue().equals(x.getValue()))) &&
+                                record.getSegmentsToSeal().stream().allMatch(segmentsToSeal::contains))) {
+                            log.debug("scale conflict, another scale operation is ongoing");
+                            throw new ScaleOperationExceptions.ScaleConflictException();
                         }
+                        return CompletableFuture.completedFuture(record);
+                    } else {
+                        // if state is SCALING and epoch transition record does not exist, reset the state back to ACTIVE
+                        return resetStateConditionally(State.SCALING)
+                                .thenCompose(v -> {
+                                    if (runOnlyIfStarted) {
+                                        log.info("scale not started, retry later.");
+                                        throw new TaskExceptions.StartException("Scale not started yet.");
+                                    }
 
-                        log.info("Scale {}/{} for segments started. Creating new segments. SegmentsToSeal {}", scope, name, sealedSegments);
-                        // fresh run
-                        // Ensure that segment.creation time is monotonically increasing after each new scale.
-                        // because scale time could be supplied by a controller with a skewed clock, we should:
-                        // take max(scaleTime, lastScaleTime + 1, System.currentTimeMillis)
-                        long lastScaleTime = HistoryRecord.readLatestRecord(historyTable.getData(), true).map(HistoryRecord::getScaleTime).orElse(0L);
-
-                        long scaleEventTime = Math.max(System.currentTimeMillis(), scaleTimestamp);
+                                    // check input is valid and satisfies preconditions
+                                    if (!TableHelper.canScaleFor(segmentsToSeal, historyTable.getData())) {
+                                        // invalid input, log and ignore
+                                        log.warn("scale precondition failed {}", segmentsToSeal);
+                                        throw new ScaleOperationExceptions.ScalePreConditionFailureException();
+                                    }
 
-                        long segmentCreationTimestamp = Math.max(scaleEventTime, lastScaleTime + 1);
+                                    EpochTransitionRecord epochTransition = TableHelper.computeEpochTransition(
+                                            historyTable.getData(), segmentTable.getData(), segmentsToSeal, newRanges, scaleTimestamp);
 
-                        return scaleCreateNewSegments(newRanges, segmentCreationTimestamp, segmentTable, activeEpoch);
+                                    return createEpochTransitionNode(epochTransition.toByteArray())
```

Comment:
Is it possible that an epoch transition node is created between the call to get the epoch transition information in line 475 and this line?

Reply:
yes.. it is possible. 
The create should fail with Data_Exists and start scale method should fail with ScaleConflictException.
I will throw the appropriate exception.

Reply:
OK, is it possible that:

1- We execute line 175 and find no node
2- Concurrently, another scaling even creates the transition epoch znode
3- The concurrent scaling event  finishes and deletes the epoch transition znode before we hit this line

In this case, this present scaling event will be completely unaware that a scaling event has been processed in parallel, which I think is problematic.

Reply:
Yes, that is a possibility. 
However, do note that the scale workflow against that EpochTransition entry will eventually fail while trying to update the actual metadata with either `IllegalArgumentException` or `ScaleConditionInvalidException` at subsequent steps during scale.
The problem is that we will end up telling the caller that scale has been started. 
And across those failures we dont reset the state as it is an unexpected situation.  
We need to reset it to ACTIVE. I will make the fix. 
Good catch. Thanks.